File: snail_pipes/URLInput.py
from urllib.parse import urljoin

import requests
from bs4 import BeautifulSoup
import re


def process(url):
    try:
        headers = {
            'User-Agent': 'Python',
        }
        response = requests.get(url, headers=headers)
        source_code = response.text
        etag = response.headers.get("ETag")
        if etag is not None:
            etag = etag.replace("\"", "")
        soup = BeautifulSoup(source_code, 'html.parser')

        text = soup.get_text(separator=' ')
        wordlist = re.split('[^a-zA-Z]', text.lower())

        # --- Create an anchor list ---
        anchorlist = []
        for a in soup.find_all('a', href=True):
            anchorlist.append(a['href'])
        return ProcessResult(wordlist, anchorlist, etag)
    except:
        return ProcessResult([], [], "")

class ProcessResult:
    def __init__(self, wordlist, anchorlist, etag):
        self.wordlist = wordlist
        self.anchorlist = anchorlist
        self.etag = etag

    def get_anchor_urls(self, base_url):
        urls_to_process = []
        for anchor in self.anchorlist:
            full_url = urljoin(base_url, anchor)
            urls_to_process.append(full_url)
        return urls_to_process

File: src/main/java/search/URLInput.java
package search;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.IOException;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

/**
 * Port of snail_pipes.URLInput.
 */
public class URLInput {

    private static final Pattern WORD_SPLIT = Pattern.compile("[^a-zA-Z]");
    private static final HttpClient HTTP_CLIENT = HttpClient.newBuilder()
            .followRedirects(HttpClient.Redirect.NORMAL)
            .build();

    public static ProcessResult process(String url) {
        try {
            HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(url))
                    .header("User-Agent", UserAgent.getUserAgent())
                    .GET()
                    .build();

            HttpResponse<String> response = HTTP_CLIENT.send(request, HttpResponse.BodyHandlers.ofString());
            String sourceCode = response.body();
            String etag = response.headers().firstValue("ETag").orElse(null);
            if (etag != null) {
                etag = etag.replace("\"", "");
            }

            Document soup = Jsoup.parse(sourceCode);
            String text = soup.text();
            List<String> wordlist = Arrays.stream(WORD_SPLIT.split(text.toLowerCase()))
                    .filter(w -> !w.isEmpty())
                    .collect(Collectors.toList());

            List<String> anchorlist = new ArrayList<>();
            Elements links = soup.select("a[href]");
            for (Element a : links) {
                String href = a.attr("href");
                if (href != null && !href.isEmpty()) {
                    anchorlist.add(href);
                }
            }
            return new ProcessResult(wordlist, anchorlist, etag != null ? etag : "");
        } catch (IOException | InterruptedException e) {
            return new ProcessResult(List.of(), List.of(), "");
        }
    }
}
